allennlp/data/semparse/type_declarations/type_declaration.py

```diff
@@ -254,3 +257,128 @@ def make_VariableExpression(self, name):
+    <#1,#1> for example, this may substitute the placeholders with different basic types. In that case, you'd
+    want to use ``_substitute_placeholder_type`` instead.
+    """
+    if _type == ANY_TYPE:
+        return basic_types
+    if isinstance(_type, (BasicType, PlaceholderType)):
+        return set([_type])
+    substitutions = set()
+    for first_type in _substitute_any_type(_type.first, basic_types):
+        for second_type in _substitute_any_type(_type.second, basic_types):
+            substitutions.add(ComplexType(first_type, second_type))
+    return substitutions
+
+
+def _substitute_placeholder_type(_type: Type, basic_type: BasicType) -> Type:
+    """
+    Takes a type with placeholders, and a basic type and substitutes all occurrences of the placeholder with
+    that type.
+    """
+    # TODO (pradeep): This assumes there's just one placeholder in the type. So this doesn't work with
+    # ``reverse`` yet, which has two placeholders.
+    if _type == ANY_TYPE:
+        return basic_type
+    if isinstance(_type, BasicType):
+        return _type
+    return ComplexType(_substitute_placeholder_type(_type.first, basic_type),
+                       _substitute_placeholder_type(_type.second, basic_type))
+
+
+def _make_production_string(source: Type, target: Union[List[Type], Type]) -> str:
+    return "%s -> %s" % (str(source), str(target))
+
+
+def _get_complex_type_productions(complex_type: ComplexType) -> List[Tuple[str, str]]:
+    """
+    Takes a complex type without any placeholders and returns all productions that lead to it, starting
+    from the most basic return type. For example, if the complex is `<a,<<b,c>,d>>`, this gives the
+    following tuples
+    ('<<b,c>,d>', '<<b,c>,d> -> [<a,<<b,c>,d>>, a]')
+    ('d', 'd -> [<<b,c>,d>, <b,c>]')
+    """
+    all_productions = []
+    while isinstance(complex_type, ComplexType) and not complex_type == ANY_TYPE:
+        all_productions.append((str(complex_type.second), _make_production_string(complex_type.second,
+                                                                                  [complex_type,
+                                                                                   complex_type.first])))
+        complex_type = complex_type.second
+    return all_productions
+
+
+def get_valid_actions(name_mapping: Dict[str, str],
+                      type_signatures: Dict[str, Type],
+                      basic_types: Set[Type]) -> Dict[str, Set[str]]:
+    """
+    Generates all the valid actions starting from each non-terminal. For terminals of a specific type, we
+    simply add to valid actions, productions from thee types to the terminals. Among those types, we keep
+    track of all the non-basic types (i.e., function types). For those types, we infer the list of productions
+    that start from a basic type leading to them.
+    For complex types that do not contain ANY_TYPE or placeholder
```

Comment:
Do you use vim?  I can't remember if you do or not.  If you do, can you set your textwidth to 99, then run `gwap` on this paragraph?  You'd need to add blank lines above and below the paragraph, so that `gwap` works right, then remove them.

Anyway, the point is that the docstring is a little too long on some lines, and too short on others.  Vim will easily fix this for you; not sure how to do it with other editors.

Reply:
Done. I use vim.